File: leetcode/0547-number-of-provinces/solution.java
class Solution {
    public int findCircleNum(int[][] isConnected) {
        int V = isConnected.length;
        boolean[] vis = new boolean[V];
        int provinces = 0;

        //perform dfs on all nodes
        for(int i = 0; i < V; i++){
            if(!vis[i]){
                dfs(isConnected, i, vis);
                provinces++;
            }
        }

        //number of dfs calls equal number of provinces
        return provinces;
    }
    public static void dfs(int[][] adjMatrix, int u, boolean[] vis){
        vis[u] = true;
        for (int v = 0; v < adjMatrix.length; v++) {
            if (adjMatrix[u][v] == 1 && !vis[v]) {
                dfs(adjMatrix, v, vis);
            }
        }
    }
}

File: leetcode/0547-number-of-provinces/solution.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        provinces = 0

        adj = [[] for _ in range(n)]
        for i in range(n):
            for j in range(n):
                if isConnected[i][j] == 1:
                    adj[i].append(j)
        
        visited = [False] * n

        for i in range(n):
            if not visited[i]:
                self.dfs(i, adj, visited)
                provinces += 1
        
        return provinces
    
    def dfs(self, u: int, adjList: List[List[int]], visited: List[bool]):
        visited[u] = True

        edges = adjList[u]

        for v in edges:
            if not visited[v]:
                self.dfs(v, adjList, visited)
        
        return
